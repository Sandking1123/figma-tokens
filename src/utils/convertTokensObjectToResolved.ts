import {mergeTokenGroups, resolveTokenValues} from '@/plugin/tokenHelpers';
import convertTokensToGroupedObject from './convertTokensToGroupedObject';
import parseTokenValues from './parseTokenValues';

// Takes Figma Tokens input, resolves all aliases while respecting user's theme choice and outputs an object with resolved tokens, ready to be consumed by style dictionary.
export default function convertTokensObjectToResolved(tokens, usedSets = [], excludedSets = []) {
    // Parse tokens into array structure
    const parsed = parseTokenValues(tokens);
    // Merge to one giant array
    const merged = mergeTokenGroups(parsed, usedSets);
    // Resolve aliases
    const resolved = resolveTokenValues(merged);
    // Group back into one object
    const object = convertTokensToGroupedObject(resolved, excludedSets);
    return object;
}
