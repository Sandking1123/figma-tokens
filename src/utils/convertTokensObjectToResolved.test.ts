import convertTokensObjectToResolved from './convertTokensObjectToResolved';

describe('convertTokensObjectToResolved', () => {
    it('converts object-like unresolved tokens to resolved object', () => {
        const tokens = {
            options: {
                colors: {
                    red: {
                        value: '#ff0000',
                        type: 'color',
                    },
                    blue: {
                        value: '#0000ff',
                        type: 'color',
                    },
                },
                sizing: {
                    base: {
                        value: '2',
                        type: 'sizing',
                    },
                    scale: {
                        value: '1.5',
                        type: 'sizing',
                    },
                    small: {
                        value: '1 * {sizing.base}',
                        type: 'sizing',
                    },
                    medium: {
                        value: '{sizing.small} * {sizing.scale}',
                        type: 'sizing',
                    },
                },
            },
            theme: {
                colors: {
                    primary: {
                        value: '$colors.red',
                        type: 'color',
                    },
                },
            },
        };

        expect(convertTokensObjectToResolved(tokens)).toEqual({
            colors: {
                red: {
                    value: '#ff0000',
                    type: 'color',
                },
                blue: {
                    value: '#0000ff',
                    type: 'color',
                },
                primary: {
                    value: '#ff0000',
                    type: 'color',
                },
            },
            sizing: {
                base: {
                    value: 2,
                    type: 'sizing',
                },
                scale: {
                    value: 1.5,
                    type: 'sizing',
                },
                small: {
                    value: 2,
                    type: 'sizing',
                },
                medium: {
                    value: 3,
                    type: 'sizing',
                },
            },
        });
    });

    it('respects used sets', () => {
        const tokens = {
            global: {
                colors: {
                    white: {
                        value: '#ffffff',
                        type: 'color',
                    },
                    black: {
                        value: '#000000',
                        type: 'color',
                    },
                },
            },
            light: {
                colors: {
                    background: {
                        value: '$colors.white',
                        type: 'color',
                    },
                },
            },
            dark: {
                colors: {
                    background: {
                        value: '$colors.black',
                        type: 'color',
                    },
                },
            },
        };

        expect(convertTokensObjectToResolved(tokens, ['global', 'light'])).toEqual({
            colors: {
                white: {
                    value: '#ffffff',
                    type: 'color',
                },
                black: {
                    value: '#000000',
                    type: 'color',
                },
                background: {
                    value: '#ffffff',
                    type: 'color',
                },
            },
        });
    });
});
