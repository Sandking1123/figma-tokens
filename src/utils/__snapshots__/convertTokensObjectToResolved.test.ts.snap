// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertTokensObjectToResolved converts object-like unresolved tokens to resolved object 1`] = `
Object {
  "colors": Object {
    "blue": Object {
      "type": "color",
      "value": "#0000ff",
    },
    "primary": Object {
      "type": "color",
      "value": "#ff0000",
    },
    "red": Object {
      "type": "color",
      "value": "#ff0000",
    },
  },
  "sizing": Object {
    "base": Object {
      "type": "sizing",
      "value": 2,
    },
    "medium": Object {
      "type": "sizing",
      "value": 3,
    },
    "scale": Object {
      "type": "sizing",
      "value": 1.5,
    },
    "small": Object {
      "type": "sizing",
      "value": 2,
    },
  },
}
`;

exports[`convertTokensObjectToResolved does not expand typography when not needed 1`] = `
Object {
  "typography": Object {
    "h1": Object {
      "type": "typography",
      "value": Object {
        "fontFamily": "Roboto",
        "fontSize": 96,
        "fontWeight": "Light",
      },
    },
    "h2": Object {
      "type": "typography",
      "value": Object {
        "fontFamily": "Roboto",
        "fontSize": 60,
        "fontWeight": "Light",
      },
    },
    "h3": Object {
      "type": "typography",
      "value": Object {
        "fontFamily": "Roboto",
        "fontSize": 48,
        "fontWeight": "Light",
      },
    },
  },
}
`;

exports[`convertTokensObjectToResolved respects used sets 1`] = `
Object {
  "colors": Object {
    "background": Object {
      "type": "color",
      "value": "#ffffff",
    },
    "black": Object {
      "type": "color",
      "value": "#000000",
    },
    "white": Object {
      "type": "color",
      "value": "#ffffff",
    },
  },
}
`;
